### PR TITLE
Only share InMemoryCache across builds if the same classloader was used to load the plugin

### PR DESCRIPTION
**Pull request by jjohannes**
Follow up to #914 

Users can run into https://github.com/gradle/gradle/issues/17559 is the plugin is loaded with different classloaders in different builds. In these cases we now keep the old behavior of having separate caches for separate builds.

There is now a test checking the two different cases.

### CODE REVIEW

File: src/main/kotlin/com/autonomousapps/subplugin/ProjectPlugin.kt [hunk 1]
```diff
@@ -58,8 +58,6 @@ internal class ProjectPlugin(private val project: Project) {
     }
   }
 
-  private lateinit var inMemoryCacheProvider: Provider<InMemoryCache>
```
Comment (jjohannes) [inline src/main/kotlin/com/autonomousapps/subplugin/ProjectPlugin.kt:61]: Removed to not register too early (i.e. before all builds got the chance to apply the plugin).

Also, this additional state is not necessary. The `registerIfAbsent` of Gradle keeps care of holding the already registered service and should be optimized for being called many times.

File: src/functionalTest/groovy/com/autonomousapps/jvm/projects/IncludedBuildWithDivergingPluginClasspathsProject.groovy [hunk 2]
```diff
@@ -0,0 +1,43 @@
...
+  IncludedBuildWithDivergingPluginClasspathsProject(boolean divergingPluginClasspaths) {
+    this.gradleProject = build(divergingPluginClasspaths)
+  }
+
+  private GradleProject build(boolean divergingPluginClasspaths) {
+    def builder = newGradleProjectBuilder()
+    def printServiceObject = """\
+      afterEvaluate { // needs 'afterEvaluate' because plugin code also uses 'afterEvaluate'
+        tasks.named('explodeJarMain') {
+          doLast { println(inMemoryCache.get()) }
```
Comment (autonomousapps) [inline src/functionalTest/groovy/com/autonomousapps/jvm/projects/IncludedBuildWithDivergingPluginClasspathsProject.groovy:20]: clever 😅

File: src/main/kotlin/com/autonomousapps/subplugin/RootPlugin.kt [hunk 3]
```diff
@@ -18,6 +18,8 @@ import org.gradle.api.artifacts.Configuration
 import org.gradle.kotlin.dsl.apply
 import org.gradle.kotlin.dsl.register
 
+const val DEPENDENCY_ANALYSIS_PLUGIN = "com.autonomousapps.dependency-analysis"
```
Comment (autonomousapps) [inline src/main/kotlin/com/autonomousapps/subplugin/RootPlugin.kt:21]: Do we want this public, or is `internal` better?

**Review by autonomousapps — APPROVED** [review]
Left one question about visibility.

File: src/main/kotlin/com/autonomousapps/subplugin/RootPlugin.kt (hunk 3 above)
Comment (jjohannes) [reply src/main/kotlin/com/autonomousapps/subplugin/RootPlugin.kt]: Yes `internal` is good. Thanks.

It's not in my repertoire as "doing a lot of Java" programmer. I tend to forget about it. 😅

**Review by autonomousapps — APPROVED** [review]

**Pull request merged**